Bluetooth: Controller: Central and Peripheral multiple CIS support

cvinayak opened the pull request:
Implementation of preliminary Central with multiple CIS
create support in Lower Link Layer.

Updated nRF5 Series LLL implementation to support multiple
peripheral CIS establishment from peer central devices.

Add BabbleSim test to verify creation of Central multiple
CIS in a CIG, connect ACL to nine peripherals and create
CIS to each peripheral.

Add BabbleSim test to verify creation of peripheral multiple
CIS in a CIG, connect ACL to one peripherals and create nine
CIS to that peripheral device.

Fixes #55264.
Fixes #55265.
Fixes #56232.
Fixes #55266.
Fixes #56985.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Preliminary Central multiple CIS support

Implementation of preliminary Central with multiple CIS
create support in Lower Link Layer. Uses a simple CIS
offset calculation that works when ACL are created with
connection interval that is multiple of CIG ISO interval.
Each ACLs are offset using BT_CTLR_CENTRAL_SPACING hence
having the same CIG sync delay values for all the ACLs.
Advanced Scheduling to setup CIS in any order of ACL and
CIS combination, ACL placed anywhere in timeline and CIS
creation will come therein will be implemented in later
Pull Requests.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
tests: Bluetooth: bsim: Add peripheral multiple CIS in a CIG test

Add BabbleSim test to verify creation of peripheral multiple
CIS in a CIG, connect ACL to one peripherals and create nine
CIS to that peripheral device.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
tests: Bluetooth: bsim: Add Central multiple CIS in a CIG test

Add BabbleSim test to verify creation of Central multiple
CIS in a CIG, connect ACL to nine peripherals and create
CIS to each peripheral.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Include remainder when calculating cis_offset

Include ticker expire remainder value when calculating the
cis_offset to acheive fine timing in scheduling the CIG
radio events relative to its associated ACL radio events.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Ticker first expire with remainder

To acheive microsecond or fine timing for first expire
timeout, pass remainder value of the coarse tick to a new
ticker_start_us() interface. The new interface is available
with BT_TICKER_EXT feature enabled.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Use ticker_start_us for Central and ISO setup

Use the ticker_start_us() interface for Central and Central
ISO setup to acheive fine timing for the first expire
timeout.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Use Conn ISO prepare latency to check timeout

Use accumulated connected ISO LLL prepare callback latencies
to check supervision timeout.

Also use the accumulated connected ISO LLL prepare callback
latencies to adjust SN and NESN values.

Connection establishment can cause short prepare callbacks
inserted before Connected ISO group prepare callbacks, that
cause the Connected ISO group prepare callbacks being
cancelled while being in prepare pipeline. Hence, it is
necessary to accumulate prepare callback latencies wherein
ticker callbacks are executed but prepare callbacks are
cancelled in the prepare pipeline.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Add BT_TICKER_EXT_EXPIRE_INFO Kconfig

Add BT_TICKER_EXT_EXPIRE_INFO Kconfig.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Revert "Bluetooth: controller: Use new ticker feature to get offsets"

This reverts commit 714d74b77b1f4bad9df427511a6a1fb1d6b74b6e.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Extended Advertising using ticker expire info

This is re-introduction of Extended, Periodic Advertising
and Broadcast ISO using ticker expire info feature to get
the aux, sync and BIG offsets.

Relates to commit 714d74b77b1f ("Bluetooth: controller: Use
new ticker feature to get offsets").

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix implementation of CIS Null PDU Indicator use

Fix implementation of CIS Null PDU Indicator use to meet the
BT Specification statement, that, if the source of the data
fails to provide a payload in time for a CIS subevent, then
the link layer shall transmit a CIS Null PDU instead.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Central multiple CIS using advanced scheduling

Central multiple CIS create implementation using advanced
scheduling to find the active CIG events and calculate the
CIS offset to be used in the LL_CIS_REQ PDU.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
tests: Bluetooth: bsim: Central multiple CIS with ACL group

Add BabbleSim test to verify creation of Central multiple
CIS in a CIG, connect ACL to three peripherals in a group
in the timeline and create CIS to each peripheral.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on tests/bsim/bluetooth/ll/cis/src/main.c:
```diff
@@ -0,0 +1,801 @@
+		.test_descr = "Peripheral ISO 4",
+		.test_post_init_f = test_cis_init,
+		.test_tick_f = test_cis_tick,
+		.test_main_f = test_cis_peripheral4,
+	},
+	{
+		.test_id = "peripheral5",
+		.test_descr = "Peripheral ISO 5",
+		.test_post_init_f = test_cis_init,
+		.test_tick_f = test_cis_tick,
+		.test_main_f = test_cis_peripheral5,
+	},
+	{
+		.test_id = "peripheral6",
+		.test_descr = "Peripheral ISO 6",
+		.test_post_init_f = test_cis_init,
+		.test_tick_f = test_cis_tick,
+		.test_main_f = test_cis_peripheral6,
+	},
+	{
+		.test_id = "peripheral7",
+		.test_descr = "Peripheral ISO 7",
+		.test_post_init_f = test_cis_init,
+		.test_tick_f = test_cis_tick,
+		.test_main_f = test_cis_peripheral7,
+	},
+	{
+		.test_id = "peripheral8",
+		.test_descr = "Peripheral ISO 8",
+		.test_post_init_f = test_cis_init,
+		.test_tick_f = test_cis_tick,
+		.test_main_f = test_cis_peripheral8,
+	},
+	{
+		.test_id = "peripheral9",
+		.test_descr = "Peripheral ISO 9",
+		.test_post_init_f = test_cis_init,
+		.test_tick_f = test_cis_tick,
+		.test_main_f = test_cis_peripheral9,
+	},
```
I don't see any reason to define multiple `test_id`s for the peripheral, as they just call the same function anyhow. You can have multiple devices run the same `test_id`. 

If you want to have anything specific for them based on their number (1, 2, 3, ...) you could also use test arguments instead. 

So I suggest to only define a single `peripheral` test ID, and just use that for all the peripherals.

Thalley commented on tests/bsim/bluetooth/ll/cis/tests_scripts/connected_iso.sh:
```diff
@@ -0,0 +1,63 @@
+verbosity_level=2
+process_ids=""; exit_code=0
+
+function Execute(){
+  if [ ! -f $1 ]; then
+    echo -e "  \e[91m`pwd`/`basename $1` cannot be found (did you forget to\
+ compile it?)\e[39m"
+    exit 1
+  fi
+  timeout 120 $@ & process_ids="$process_ids $!"
+}
```
The babblesim test scripts were recently refactored to not define `Execute` themselves, but rather get it from a common file
```c
source ${ZEPHYR_BASE}/tests/bsim/sh_common.source
```

cvinayak replied:
Done

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_central_iso.c:
```diff
@@ -684,11 +694,17 @@ uint8_t ull_central_iso_setup(uint16_t cis_handle,
 	cis->offset = cis_offset;
 
 #else /* !CONFIG_BT_CTLR_JIT_SCHEDULING */
+#if (CONFIG_BT_CTLR_CENTRAL_SPACING > 0)
```
IS_ENABLED?

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_central_iso.c:
```diff
@@ -725,32 +743,166 @@ uint16_t ull_central_iso_cis_offset_get(uint16_t cis_handle, uint32_t *cis_offse
+		return 0;
+	}
+
+#if (CONFIG_BT_CTLR_CENTRAL_SPACING == 0)
```
IS_ENABLED?

cvinayak replied:
Done

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_central_iso.c:
```diff
@@ -725,32 +743,166 @@ uint16_t ull_central_iso_cis_offset_get(uint16_t cis_handle, uint32_t *cis_offse
+	id = TICKER_NULL;
+	ticks_to_expire = 0U;
+	ticks_current = 0U;
+	retry = 4U;
```
What's the rationale behind retrying 4 times? Could you add that as a comment for this value?

cvinayak replied:
Added comment.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_sched.c:
```diff
@@ -352,6 +296,119 @@ static void after_cen_offset_get(uint16_t conn_interval, uint32_t ticks_slot,
+
+		if (false) {
+
+#if defined(CONFIG_BT_BROADCASTER)
```
Can some of the `#if defined` here be changed into `IS_ENABLED`?

cvinayak replied:
There are defines used in `IN_RANGE` that are conditional hence, not able to use `IS_ENABLED` .

cvinayak committed:
tests: bsim: Bluetooth: Add BT_TICKER_EXT_EXPIRE_INFO coverage

Extend tests to cover use of BT_TICKER_EXT_EXPIRE_INFO
feature.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Preliminary multiple peripheral CIS support

Updated nRF5 Series LLL implementation to support multiple
peripheral CIS establishment from peer central devices.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Group depends on together in Kconfig file

Group depends on clauses together in the Kconfig.ll_sw_split
file.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix BT_TICKER_EXT BT_TICKER_LOW_LAT dependency

BT_TICKER_EXT is not supported with BT_TICKER_LOW_LAT.
Remove redundant select of BT_TICKER_EXT for
BT_CTLR_ADV_EXT.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix BT_TICKER_NEXT_SLOT_GET select

Fix BT_TICKER_NEXT_SLOT_GET select, to depend on
BT_CTLR_SCHED_ADVANCED.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Remove redundant Kconfig range and default

Remove redundant range and default for BT_CTLR_ADV_AUX_SET
when supporting BT_CTLR_ADV_PERIODIC.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Remainder and Lazy depends on match get

Ticker remainder and lazy get depends on match get interface
provided by BT_TICKER_NEXT_SLOT_GET_MATCH.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_conn_iso.c:
```diff
@@ -708,8 +708,9 @@ static void ticker_op_cb(uint32_t status, void *param)
-void ull_conn_iso_start(struct ll_conn *conn, uint32_t ticks_at_expire,
-			uint16_t cis_handle, uint16_t instant_latency)
+void ull_conn_iso_start(struct ll_conn *conn, uint16_t cis_handle,
+			uint32_t ticks_at_expire, uint32_t remainder,
```
`remainder` is a value in femto seconds (in nRF) and in vendor specific unit. `remainder_us` in microseconds is obtained using `hal_ticker_remove_jitter()` HAL interface.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_conn_types.h:
```diff
@@ -67,6 +67,7 @@ struct llcp_struct {
 	/* Prepare parameters */
 	struct {
 		uint32_t ticks_at_expire;
+		uint32_t remainder;
```
This is in vendor specific unit.

Thalley replied:
Should that perhaps be documented in the struct then?

cvinayak commented on subsys/bluetooth/controller/ticker/ticker.h:
```diff
@@ -64,16 +64,24 @@
+#define TICKER_USER_OP_T_SIZE   52
+#else /* !CONFIG_BT_TICKER_REMAINDER */
 #define TICKER_USER_OP_T_SIZE   48
-#endif /* CONFIG_BT_TICKER_SLOT_AGNOSTIC */
-#else
+#endif /* !CONFIG_BT_TICKER_REMAINDER */
+#endif /* !CONFIG_BT_TICKER_SLOT_AGNOSTIC */
+#else /* !CONFIG_BT_TICKER_EXT */
 #if defined(CONFIG_BT_TICKER_SLOT_AGNOSTIC)
 #define TICKER_USER_OP_T_SIZE   40
-#else
+#else /* !CONFIG_BT_TICKER_SLOT_AGNOSTIC */
+#if defined(CONFIG_BT_TICKER_REMAINDER)
+#define TICKER_USER_OP_T_SIZE   48
+#else /* !CONFIG_BT_TICKER_REMAINDER */
 #define TICKER_USER_OP_T_SIZE   44
-#endif /* CONFIG_BT_TICKER_SLOT_AGNOSTIC */
-#endif /* CONFIG_BT_TICKER_EXT */
+#endif /* !CONFIG_BT_TICKER_REMAINDER */
+#endif /* !CONFIG_BT_TICKER_SLOT_AGNOSTIC */
+#endif /* !CONFIG_BT_TICKER_EXT */
```
These are manually calculated struct sizes, and ticker.c file has build asserts to verify them. For historical reason, user of the ticker module (in case prebuilt library) was to supply the memory at runtime and use these size to allocate memory in the upper layers.

cvinayak committed:
tests: bsim: Bluetooth: Review rework changes for multiple CIS support

Assorted review rework changes for multiple CIS support test
implementation commits.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Review rework changes for multiple CIS support

Assorted review rework changes for multiple CIS support
implementation commits.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_central_iso.c:
```diff
@@ -143,6 +171,17 @@ static int prepare_cb(struct lll_prepare_param *p)
 
+	/* Store the current event latency */
+	cig_lll->latency_event = cig_lll->latency_prepare;
+	lazy = cig_lll->latency_prepare - 1U;
```
`latency_prepare` can never be `0` here?

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_central_iso.c:
```diff
@@ -485,11 +531,135 @@ static void isr_tx(void *param)
+		uint32_t start_us;
+		memq_link_t *link;
+		uint8_t bn;
+
+
```
```suggestion

```

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_central_iso.c:
```diff
@@ -143,6 +171,17 @@ static int prepare_cb(struct lll_prepare_param *p)
+	/* Reset accumulated latencies */
+	cig_lll->latency_prepare = 0U;
+
+	/* Adjust the SN and NESN for skipped CIG events */
+	cis_lll->sn += cis_lll->tx.bn * lazy;
+	cis_lll->nesn += cis_lll->rx.bn * lazy;
```
So `sn` and `nesn` are 1 bits each, and `lazy` is a `uint16_t`. 

Is this designed to specifically overflow?

Thalley replied:
Alright, please add a comment that states this. 

Could perhaps also be made more clear by doing `&= 1` as you just want the lowest bit, right?

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_peripheral_iso.c:
```diff
@@ -736,9 +753,64 @@ static void isr_rx(void *param)
+	subevent_us = radio_tmr_aa_restore();
+	subevent_us += cis_lll->offset - cis_offset_first +
+		       (cis_lll->sub_interval * se_curr);
+	subevent_us -= addr_us_get(cis_lll->rx.phy);
+
+#if defined(CONFIG_BT_CTLR_PHY)
+	subevent_us -= radio_rx_ready_delay_get(cis_lll->rx.phy,
+						PHY_FLAGS_S8);
+	subevent_us -= radio_rx_chain_delay_get(cis_lll->rx.phy,
+						PHY_FLAGS_S8);
+#else /* !CONFIG_BT_CTLR_PHY */
+	subevent_us -= radio_rx_ready_delay_get(0U, 0U);
+	subevent_us -= radio_rx_chain_delay_get(0U, 0U);
+#endif /* !CONFIG_BT_CTLR_PHY */
```
No risk of underflow?

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_iso.c:
```diff
@@ -710,6 +714,63 @@ uint8_t ll_adv_iso_by_hci_handle_new(uint8_t hci_handle, uint8_t *handle)
+	bi->payload_count_framing[1] = payload_count >> 8;
+	bi->payload_count_framing[2] = payload_count >> 16;
+	bi->payload_count_framing[3] = payload_count >> 24;
+	bi->payload_count_framing[4] &= ~0x7F;
+	bi->payload_count_framing[4] |= (payload_count >> 32) & 0x7F;
```
A a reader this still looks suspicious. I'd suggest to add a comment about it, or define a function/macro to do this so that becomes more clear

cvinayak replied:
Added comment.

Thalley commented on tests/bsim/bluetooth/ll/cis/tests_scripts/connected_iso_acl_group.sh:
```diff
@@ -0,0 +1,48 @@
+verbosity_level=2
+process_ids=""; exit_code=0
+
+function Execute(){
+  if [ ! -f $1 ]; then
+    echo -e "  \e[91m`pwd`/`basename $1` cannot be found (did you forget to\
+ compile it?)\e[39m"
+    exit 1
+  fi
+  timeout 120 $@ & process_ids="$process_ids $!"
+}
```
Ditto with using the new file structure

cvinayak replied:
Done

Thalley commented on tests/bsim/bluetooth/ll/cis/tests_scripts/connected_iso_peripheral_cis.sh:
```diff
@@ -0,0 +1,38 @@
+verbosity_level=2
+process_ids=""; exit_code=0
+
+function Execute(){
+  if [ ! -f $1 ]; then
+    echo -e "  \e[91m`pwd`/`basename $1` cannot be found (did you forget to\
+ compile it?)\e[39m"
+    exit 1
+  fi
+  timeout 120 $@ & process_ids="$process_ids $!"
+}
```
Ditto

cvinayak replied:
Done

Thalley reviewed (changes requested):
A few more comments

cvinayak committed:
tests: bsim: Bluetooth: Test script clean up

Test script clean up related to common code moved in
commit add1397e4a8d ("tests/bsim/bluetooth: Tests scripts
refactor").

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley reviewed (approved):
All my comments have been addressed, and the changes look sane to me
